vscode/src/context/remote-repo-searcher.ts

```diff
@@ -0,0 +1,261 @@
+     * Repeated searches with the same query and repo list are cached, so
+     * accessing subsequent pages is fast.
+     *
+     * @param query the fuzzy search query, if any.
+     * @param first the number of results to retrieve.
+     * @param after the repository ID of the last repo of the previous page.
+     * @returns a list of repositories with their position in the complete
+     * filtered list.
+     */
+    public list(query: string | undefined, first: number, after?: string): RemoteRepoListResult {
+        this.ensureCompleteOrFetching()
+        // If we haven't cached this result, execute the query.
+        if (query !== this.cachedQuery) {
+            this.cachedQuery = query
+            this.updateCachedResult(query)
+        }
+        // Find the start index, if any.
+        let start: number
+        if (after) {
+            const afterIndex = this.cachedResult.findIndex(repo => repo.id === after)
+            if (afterIndex === -1) {
+                // The after ID doesn't exist in the list, so there's nothing "after" that.
+                return {
+                    repos: [],
+                    startIndex: -1,
+                    count: this.cachedResult.length,
+                    state: this.fetcher.state,
+                    lastError: this.fetcher.lastError,
+                }
+            }
+            start = afterIndex + 1
+        } else {
+            start = 0
+        }
+        // Return the subset of results
+        return {
+            repos: this.cachedResult.slice(start, start + first),
+            startIndex: start,
+            count: this.cachedResult.length,
+            state: this.fetcher.state,
+            lastError: this.fetcher.lastError,
+        }
+    }
+
+    private updateCachedResult(query: string | undefined): void {
+        if (!query) {
+            // No query, so we return all repos in fetch order.
+            this.cachedResult = this.fetcher.repositories
+            return
+        }
+
+        if (!this.fuzzyIndex || !this.fuzzyTargetRepoMap) {
+            // Update the fuzzy search index.
+            // TODO: When speed is necessary, improve this by only indexing
+            // new/deleting removed repositories.
+            this.fuzzyIndex = []
+            this.fuzzyTargetRepoMap = new Map()
+            for (const repo of this.fetcher.repositories) {
+                const prepared = fuzzysort.prepare(repo.name)
+                this.fuzzyTargetRepoMap.set(prepared.target, repo)
+                this.fuzzyIndex.push(prepared)
+            }
+        }
+
+        // Do fuzzy search.
+        // TODO: When speed is necessary, improve this by searching the changed
+        // repositories and merging the result with the cached results, if any.
+        // Could also limit the result set and cap the maximum size of any list.
+        const result = fuzzysort.go(query, this.fuzzyIndex, {
+            threshold: -1000,
```

Comment:
```suggestion
            limit: 20,
            threshold: -1000,
```
would it make sense to limit the results for UI? I don't imagine people would want to go through a long list but that's just me, so feel free to ignore this 😄